mosaic-runtime/src/commonMain/kotlin/com/jakewharton/mosaic/mosaic.kt

```diff
@@ -30,50 +33,45 @@ import kotlinx.coroutines.launch
  */
 private const val debugOutput = false
 
-internal fun mosaicNodes(content: @Composable () -> Unit): MosaicNode {
-	val clock = BroadcastFrameClock()
-	val job = Job()
-	val composeContext = clock + job
-
-	val rootNode = createRootNode()
-	val recomposer = Recomposer(composeContext)
-	val composition = Composition(MosaicNodeApplier(rootNode), recomposer)
-
-	composition.setContent(content)
-
-	job.cancel()
-	composition.dispose()
-
-	return rootNode
-}
-
-public fun renderMosaic(content: @Composable () -> Unit): String {
-	val rootNode = mosaicNodes(content)
-	val render = AnsiRendering().render(rootNode)
-	return render.toString()
-}
-
 public suspend fun runMosaic(content: @Composable () -> Unit) {
 	coroutineScope {
-		val mosaicComposition = MosaicComposition(coroutineScope = this, onRender = ::platformDisplay)
+		val mosaicComposition = RenderingMosaicComposition(
+			coroutineScope = this,
+			onRender = ::platformDisplay,
+		)
 		mosaicComposition.setContent(content)
 		mosaicComposition.awaitComplete()
 	}
 }
 
-internal class MosaicComposition(
+public suspend fun renderMosaic(content: @Composable () -> Unit): String {
```

Comment:
Since the initial composition is synchronous, the idea was that this function could always be synchronous, too, and that no effects would run at all. I would strongly prefer to keep a design where we can retain this behavior. I believe we can create our own `CoroutineScope`, rely on `onRender` being called synchronously, and then cancel everything to keep it that way.